[jit][await] Add await then fns

IvanKobzarev opened the pull request:
Introducing _awaitable_then operator to add continuation function to the awaitable.

For example:
```
        def delayed(x: Tensor) -> Tensor:
            return -1 * x

        def then(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 5 * aw_out

        def then2(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 4 * aw_out

        def main(x: Tensor) -> Tensor:
            aw = torch.jit._awaitable(delayed, x)
            torch.jit._awaitable_then(then, aw)
            torch.jit._awaitable_then(then2, aw)
            return torch.jit._awaitable_wait(aw)
```
`awaitable_wait(aw)` is `then2(then(delayed(x)))`

Inside ivalue `then` continuations are stored as std::function which will be called on the result when `wait` is called.

IR:

torch.jit._awaitable_then first will emit `awaitableThenClosure` (the same as fork, awaitable), 
but that closure will have two arguments: `Await[T]` and fake argument of `T` type, produced by no-op operator `awaitable_then_input`.
This fake argument in runtime will become the result of Await delayed function.
We need this fake argument to generate valid call to `then` function without having that Value at the moment of generation.

(This happens in `torch/csrc/jit/frontend/ir_emitter.cpp` and `torch/csrc/jit/passes/inline_forked_closures.cpp` where `prim::awaitable_then_input` is added)


New instruction `AWAITABLE_THEN` is added, which adds continuation to Await on the stack (the first argument of Closure).


python_print:

As awaitable_then_input is a fake argument - it is ignored for printing.
awaitable_then is printed exactly the same way as python is specified.

Python:
```
def gap(x: Tensor):
    return torch.relu(x)

def delayed(x: Tensor) -> Tensor:
    return -1 * x

def continuation_fn(aw: Await[Tensor], x: Tensor) -> Tensor:
    return 5 * x

def main(x: Tensor) -> Tensor:
    aw = torch.jit._awaitable(delayed, x)
    z = gap(x)
    torch.jit._awaitable_then(continuation_fn, aw)

    y = torch.jit._awaitable_wait(aw)
    return x + y + z
```

IR flow:

```
graph(%self : __torch__.PlaceholderModule,
      %x.1 : Tensor):
  %41 : Tensor = prim::Uninitialized()
  %40 : bool = prim::Uninitialized()
  %38 : bool = prim::Constant[value=1]()
  %39 : bool = prim::Constant[value=0]()
  %37 : bool = prim::Uninitialized()
  %35 : bool = prim::Constant[value=1]()
  %36 : bool = prim::Constant[value=0]()
  %32 : bool = prim::Constant[value=0]()
  %2 : NoneType = prim::Closure_0()
  %42 : (Tensor) = prim::TupleConstruct(%x.1)
  %43 : (NoneType, (Tensor)) = prim::TupleConstruct(%2, %42)
  %aw.1 : Await(Tensor) = prim::awaitableClosure(%43)
  %10 : Function = prim::Constant[name="gap"]()
  %z.1 : Tensor = prim::CallFunction(%10, %x.1)
  %12 : NoneType = prim::Closure_1()
  %44 : () = prim::TupleConstruct()
  %45 : (NoneType, ()) = prim::TupleConstruct(%12, %44)
   = prim::awaitableThenClosure(%45, %aw.1)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1)
  %26 : int = prim::Constant[value=1]()
  %27 : Tensor = aten::add(%x.1, %y.1, %26)
  %29 : int = prim::Constant[value=1]()
  %30 : Tensor = aten::add(%27, %z.1, %29)
  return (%30)
with prim::Closure_0 = graph(%context : (Tensor)):
  %2 : Function = prim::Constant[name="delayed"]()
  %x.1 : Tensor = prim::TupleUnpack(%context)
  %3 : Tensor = prim::CallFunction(%2, %x.1) # :0:0
  return (%3)
with prim::Closure_1 = graph(%context : (),
      %aw0.1 : Await(Tensor),
      %x0.1 : Tensor):
  %4 : Function = prim::Constant[name="continuation_fn"]()
  %5 : Tensor = prim::CallFunction(%4, %aw0.1, %x0.1) # :0:0
  return (%5)
```
After applying `inline_forked_closures.cpp`
```
graph(%x.1 : Tensor):
  %19 : int = prim::Constant[value=1]()
  %7 : Function = prim::Constant[name="gap"]()
  %aw.1 : Await(Tensor) = prim::awaitable_0(%x.1)
  %z.1 : Tensor = prim::CallFunction(%7, %x.1)
  %40 : Tensor = prim::awaitable_then_input(%aw.1)
   = prim::awaitable_then_1(%aw.1, %40)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1)
  %20 : Tensor = aten::add(%x.1, %y.1, %19)
  %23 : Tensor = aten::add(%20, %z.1, %19)
  return (%23)
with prim::awaitable_0 = graph(%x.1 : Tensor):
  %1 : Function = prim::Constant[name="delayed"]()
  %3 : Tensor = prim::CallFunction(%1, %x.1)
  return (%3)
with prim::awaitable_then_1 = graph(%aw.5 : Await(Tensor),
      %aw_out : Tensor):
  %3 : Function = prim::Constant[name="continuation_fn"]()
  %4 : Tensor = prim::CallFunction(%3, %aw.5, %aw_out)
  return (%4)
```

`python_print.cpp` produces:

```
def main(x: Tensor) -> Tensor:
  def __awaitable_function():
    _0 = __torch__.jit.test_await.delayed(x, )
    return _0
  aw = awaitable(__awaitable_function)
  z = __torch__.jit.test_await.gap(x, )
  def __awaitable_then_function(aw0 : Await[Tensor], x0 : Tensor):
    _1 = __torch__.jit.test_await.continuation_fn(aw0, x0, )
    return _1
  awaitable_then(__awaitable_then_function, aw)
  y = ops.prim.awaitable_wait(aw)
  return torch.add(torch.add(x, y), z)
```


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99623
* __->__ #94811

Differential Revision: [D43572964](https://our.internmc.facebook.com/intern/diff/D43572964)

IvanKobzarev committed:
[jit][await] Add await then fns

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"

Introducing _awaitable_then operator to add continuation function to the awaitable.

For example:
```
        def delayed(x: Tensor) -> Tensor:
            return -1 * x

        def then(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 5 * aw_out

        def then2(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 4 * aw_out

        def main(x: Tensor) -> Tensor:
            aw = torch.jit._awaitable(delayed, x)
            torch.jit._awaitable_then(then, aw)
            torch.jit._awaitable_then(then2, aw)
            return torch.jit._awaitable_wait(aw)
```
`awaitable_wait(aw)` is `then2(then(delayed(x)))`

Inside ivalue `then` continuations are stored as std::function which will be called on the result when `wait` is called.


IR:

torch.jit._awaitable_then first will emit awaitableThenClosure (the same as fork, awaitable), 
but that closure will have two arguments: Await[T] and fake argument of T type, produced by no-op operator awaitableThenInput.
This fake argument in runtime will become the result of Await delayed function.
We need this fake argument to generate valid call to `then` function without having that Value at the moment of generation.

(This happens in torch/csrc/jit/frontend/ir_emitter.cpp and torch/csrc/jit/passes/inline_forked_closures.cpp where prim::awaitable_then_input is added)


New instruction AWAITABLE_THEN is added, which adds continuation to Await on the stack (the first argument of Closure).


python_print:

As awaitable_then_input is a fake argument - it is ignored for printing.
awaitable_then is printed exactly the same way as python is specified.

Python:
```
def gap(x: Tensor):
    return torch.relu(x)

def delayed(x: Tensor) -> Tensor:
    return -1 * x

def then(aw: Await[Tensor], x: Tensor) -> Tensor:
    return 5 * x

def main(x: Tensor) -> Tensor:
    aw = torch.jit._awaitable(delayed, x)
    z = gap(x)
    torch.jit._awaitable_then(then, aw)

    y = torch.jit._awaitable_wait(aw)
    return x + y + z
```

IR flow:

```
graph(%self : __torch__.PlaceholderModule,
      %x.1 : Tensor):
  %41 : Tensor = prim::Uninitialized()
  %40 : bool = prim::Uninitialized()
  %38 : bool = prim::Constant[value=1]()
  %39 : bool = prim::Constant[value=0]()
  %37 : bool = prim::Uninitialized()
  %35 : bool = prim::Constant[value=1]()
  %36 : bool = prim::Constant[value=0]()
  %32 : bool = prim::Constant[value=0]()
  %2 : NoneType = prim::Closure_0()
  %42 : (Tensor) = prim::TupleConstruct(%x.1)
  %43 : (NoneType, (Tensor)) = prim::TupleConstruct(%2, %42)
  %aw.1 : Await(Tensor) = prim::awaitableClosure(%43) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  %10 : Function = prim::Constant[name="gap"]()
  %z.1 : Tensor = prim::CallFunction(%10, %x.1) # :0:0
  %12 : NoneType = prim::Closure_1()
  %44 : () = prim::TupleConstruct()
  %45 : (NoneType, ()) = prim::TupleConstruct(%12, %44)
   = prim::awaitableThenClosure(%45, %aw.1)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:403:16
  %26 : int = prim::Constant[value=1]()
  %27 : Tensor = aten::add(%x.1, %y.1, %26) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  %29 : int = prim::Constant[value=1]()
  %30 : Tensor = aten::add(%27, %z.1, %29) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  return (%30)
with prim::Closure_0 = graph(%context : (Tensor)):
  %2 : Function = prim::Constant[name="delayed"]()
  %x.1 : Tensor = prim::TupleUnpack(%context)
  %3 : Tensor = prim::CallFunction(%2, %x.1) # :0:0
  return (%3)
with prim::Closure_1 = graph(%context : (),
      %aw0.1 : Await(Tensor),
      %x0.1 : Tensor):
  %4 : Function = prim::Constant[name="then"]()
  %5 : Tensor = prim::CallFunction(%4, %aw0.1, %x0.1) # :0:0
  return (%5)
```
After applying inline_forked_closures.cpp
```
graph(%x.1 : Tensor):
  %19 : int = prim::Constant[value=1]()
  %7 : Function = prim::Constant[name="gap"]()
  %aw.1 : Await(Tensor) = prim::awaitable_0(%x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  %z.1 : Tensor = prim::CallFunction(%7, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:400:16
  %40 : Tensor = prim::awaitable_then_input(%aw.1)
   = prim::awaitable_then_1(%aw.1, %40)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:403:16
  %20 : Tensor = aten::add(%x.1, %y.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  %23 : Tensor = aten::add(%20, %z.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  return (%23)
with prim::awaitable_0 = graph(%x.1 : Tensor):
  %1 : Function = prim::Constant[name="delayed"]()
  %3 : Tensor = prim::CallFunction(%1, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  return (%3)
with prim::awaitable_then_1 = graph(%aw.5 : Await(Tensor),
      %aw_out : Tensor):
  %3 : Function = prim::Constant[name="then"]()
  %4 : Tensor = prim::CallFunction(%3, %aw.5, %aw_out) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:401:12
  return (%4)
```

python_print.cpp produces:

```
def main(x: Tensor) -> Tensor:
  def __awaitable_function():
    _0 = __torch__.jit.test_await.delayed(x, )
    return _0
  aw = awaitable(__awaitable_function)
  z = __torch__.jit.test_await.gap(x, )
  def __awaitable_then_function(aw0 : Await[Tensor], x0 : Tensor):
    _1 = __torch__.jit.test_await.then(aw0, x0, )
    return _1
  awaitable_then(__awaitable_then_function, aw)
  y = ops.prim.awaitable_wait(aw)
  return torch.add(torch.add(x, y), z)
```





[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43572964).

davidberard98 commented on torch/csrc/jit/passes/inline_forked_closures.cpp:
```diff
@@ -58,6 +58,56 @@ void inlineForkedClosure(Node* fork_closure, NodeKind genKind) {
+  Node* then_node = g->create(prim::awaitable_then, 0)
+                        ->insertAfter(then_closure)
+                        ->setSourceRange(then_closure->sourceRange());
+  Node* then_input_node = g->create(prim::awaitable_then_input, 1)
```
I don't really understand why you need this, can you document it here in detail? And maybe add some code pointers in other places where awaitable_the_input shows up that point to the documentation.

IvanKobzarev replied:
To add `awaitable_then` to IR - I need to specify the `jit.Value` for its input.

At the moment of `awaitable_then` call - we do not have it, as it will be the result of `await_fn`.
`fork` and `awaitable` do not have this problem as they get real IR arguments at the moment of creation.

```
graph(%x.1 : Tensor):
  %19 : int = prim::Constant[value=1]()
  %7 : Function = prim::Constant[name="gap"]()
  %aw.1 : Await(Tensor) = prim::awaitable_0(%x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  %z.1 : Tensor = prim::CallFunction(%7, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:400:16
  %40 : Tensor = prim::awaitable_then_input(%aw.1)
   = prim::awaitable_then_1(%aw.1, %40)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:403:16
  %20 : Tensor = aten::add(%x.1, %y.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  %23 : Tensor = aten::add(%20, %z.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  return (%23)
with prim::awaitable_0 = graph(%x.1 : Tensor):
  %1 : Function = prim::Constant[name="delayed"]()
  %3 : Tensor = prim::CallFunction(%1, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  return (%3)
with prim::awaitable_then_1 = graph(%aw.5 : Await(Tensor),
      %aw_out : Tensor):
  %3 : Function = prim::Constant[name="then"]()
  %4 : Tensor = prim::CallFunction(%3, %aw.5, %aw_out) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:401:12
  return (%4)
  ```
  
  If not to specify it, but just add `%aw_out` to await_then subgraph - postprocessing of IR will fail on validation that `%aw_out` gets out of nowhere ( I can not remember the particular error and its location )

IvanKobzarev replied:
Added the comment with IR example to `torch/csrc/jit/passes/inline_forked_closures.cpp`

davidberard98 commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -845,6 +847,27 @@ struct PythonPrintImpl {
+      case prim::awaitable_then_input: {
+      } break;
+      case prim::awaitable_then: {
+        auto name = genName("__awaitable_then_function");
```
TODO(dberard)

davidberard98 commented on torch/csrc/jit/passes/inline_forked_closures.cpp:
```diff
@@ -58,6 +58,56 @@ void inlineForkedClosure(Node* fork_closure, NodeKind genKind) {
+  then_node->addInput(aw);
+  then_node->addInput(fake_aw_then_input);
+
+  auto then_graph_context = then_graph->inputs().at(0);
```
can you add a comment describing that this is just unpacking the context tuple into individual inputs.

davidberard98 reviewed (commented):
only got half way through reviewing this, but I had some comments (inline) and some high level questions:

Why is the design to modify the awaitable object in place instead of returning a new awaitable? e.g. right now you do
```python
            aw = torch.jit._awaitable(delayed, x)
            torch.jit._awaitable_then(then, aw)
            torch.jit._awaitable_then(then2, aw)
```
why not do
```python
            aw = torch.jit._awaitable(delayed, x)
            aw2 = torch.jit._awaitable_then(then, aw)
            aw3 = torch.jit._awaitable_then(then2, aw2)
```
e.g. what if `then2` returns a different type than awaitable[T] returned by`then`? is that a use case that's possible?

IvanKobzarev commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -1406,6 +1406,14 @@ struct C10_EXPORT ivalue::Await final : c10::intrusive_ptr_target {
     fn_ = std::move(fn);
   }
 
+  void then(std::function<IValue(IValue)> then_fn) {
```
`await.then` is also `jit.scriptable`.
I will add it to the comments/docs

IvanKobzarev commented:
> only got half way through reviewing this, but I had some comments (inline) and some high level questions:
> 
> Why is the design to modify the awaitable object in place instead of returning a new awaitable? e.g. right now you do
> 
> ```python
>             aw = torch.jit._awaitable(delayed, x)
>             torch.jit._awaitable_then(then, aw)
>             torch.jit._awaitable_then(then2, aw)
> ```
> 
> why not do
> 
> ```python
>             aw = torch.jit._awaitable(delayed, x)
>             aw2 = torch.jit._awaitable_then(then, aw)
>             aw3 = torch.jit._awaitable_then(then2, aw2)
> ```
> 
> e.g. what if `then2` returns a different type than awaitable[T] returned by`then`? is that a use case that's possible?

Yes, `then` continuation logic can be  expressed as an awaitable of awaitable.  
For pure-functional style that will work and we do not need this change.

But there is an existing logic in torchrec that uses current `Awaitable` as a class and checks its member fields,  isinstance():
https://github.com/pytorch/torchrec/blob/main/torchrec/distributed/embedding_sharding.py#L274

Lets think about it for jit as an operator `awaitable_arg(aw: Await, idx: int) -> Any` which returns positional argument specified at the moment of Awaitable creation. (we can replace member fields and isinstance logic with that)

If to use `awaitable(awaitable())` - the external Awaitable will be different with different creation arguments and using `awaitable_arg` will be impossible to script the logic.

Using scriptable continuation will allow to keep the external object the same.

IvanKobzarev commented on test/jit/test_await.py:
```diff
@@ -384,3 +384,112 @@ def main(x: Tensor) -> Tensor:
+        def delayed(x: Tensor) -> Tensor:
+            return -1 * x
+
+        def then(aw: Await[Tensor], x: Tensor) -> Tensor:
```
Renamed to `continuation_fn`, `continuation_fn2`

davidberard98 replied:
thanks - and can you update this in the example comments? e.g. `prim::Constant[name="then"]()`. Sorry to nitpick so much, but I think some of this is unintuitive and it will be important to have clear comments for future developers.

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"

Introducing _awaitable_then operator to add continuation function to the awaitable.

For example:
```
        def delayed(x: Tensor) -> Tensor:
            return -1 * x

        def then(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 5 * aw_out

        def then2(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 4 * aw_out

        def main(x: Tensor) -> Tensor:
            aw = torch.jit._awaitable(delayed, x)
            torch.jit._awaitable_then(then, aw)
            torch.jit._awaitable_then(then2, aw)
            return torch.jit._awaitable_wait(aw)
```
`awaitable_wait(aw)` is `then2(then(delayed(x)))`

Inside ivalue `then` continuations are stored as std::function which will be called on the result when `wait` is called.

IR:

torch.jit._awaitable_then first will emit `awaitableThenClosure` (the same as fork, awaitable), 
but that closure will have two arguments: `Await[T]` and fake argument of `T` type, produced by no-op operator `awaitable_then_input`.
This fake argument in runtime will become the result of Await delayed function.
We need this fake argument to generate valid call to `then` function without having that Value at the moment of generation.

(This happens in `torch/csrc/jit/frontend/ir_emitter.cpp` and `torch/csrc/jit/passes/inline_forked_closures.cpp` where `prim::awaitable_then_input` is added)


New instruction `AWAITABLE_THEN` is added, which adds continuation to Await on the stack (the first argument of Closure).


python_print:

As awaitable_then_input is a fake argument - it is ignored for printing.
awaitable_then is printed exactly the same way as python is specified.

Python:
```
def gap(x: Tensor):
    return torch.relu(x)

def delayed(x: Tensor) -> Tensor:
    return -1 * x

def then(aw: Await[Tensor], x: Tensor) -> Tensor:
    return 5 * x

def main(x: Tensor) -> Tensor:
    aw = torch.jit._awaitable(delayed, x)
    z = gap(x)
    torch.jit._awaitable_then(then, aw)

    y = torch.jit._awaitable_wait(aw)
    return x + y + z
```

IR flow:

```
graph(%self : __torch__.PlaceholderModule,
      %x.1 : Tensor):
  %41 : Tensor = prim::Uninitialized()
  %40 : bool = prim::Uninitialized()
  %38 : bool = prim::Constant[value=1]()
  %39 : bool = prim::Constant[value=0]()
  %37 : bool = prim::Uninitialized()
  %35 : bool = prim::Constant[value=1]()
  %36 : bool = prim::Constant[value=0]()
  %32 : bool = prim::Constant[value=0]()
  %2 : NoneType = prim::Closure_0()
  %42 : (Tensor) = prim::TupleConstruct(%x.1)
  %43 : (NoneType, (Tensor)) = prim::TupleConstruct(%2, %42)
  %aw.1 : Await(Tensor) = prim::awaitableClosure(%43) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  %10 : Function = prim::Constant[name="gap"]()
  %z.1 : Tensor = prim::CallFunction(%10, %x.1) # :0:0
  %12 : NoneType = prim::Closure_1()
  %44 : () = prim::TupleConstruct()
  %45 : (NoneType, ()) = prim::TupleConstruct(%12, %44)
   = prim::awaitableThenClosure(%45, %aw.1)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:403:16
  %26 : int = prim::Constant[value=1]()
  %27 : Tensor = aten::add(%x.1, %y.1, %26) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  %29 : int = prim::Constant[value=1]()
  %30 : Tensor = aten::add(%27, %z.1, %29) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  return (%30)
with prim::Closure_0 = graph(%context : (Tensor)):
  %2 : Function = prim::Constant[name="delayed"]()
  %x.1 : Tensor = prim::TupleUnpack(%context)
  %3 : Tensor = prim::CallFunction(%2, %x.1) # :0:0
  return (%3)
with prim::Closure_1 = graph(%context : (),
      %aw0.1 : Await(Tensor),
      %x0.1 : Tensor):
  %4 : Function = prim::Constant[name="then"]()
  %5 : Tensor = prim::CallFunction(%4, %aw0.1, %x0.1) # :0:0
  return (%5)
```
After applying `inline_forked_closures.cpp`
```
graph(%x.1 : Tensor):
  %19 : int = prim::Constant[value=1]()
  %7 : Function = prim::Constant[name="gap"]()
  %aw.1 : Await(Tensor) = prim::awaitable_0(%x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  %z.1 : Tensor = prim::CallFunction(%7, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:400:16
  %40 : Tensor = prim::awaitable_then_input(%aw.1)
   = prim::awaitable_then_1(%aw.1, %40)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:403:16
  %20 : Tensor = aten::add(%x.1, %y.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  %23 : Tensor = aten::add(%20, %z.1, %19) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:404:19
  return (%23)
with prim::awaitable_0 = graph(%x.1 : Tensor):
  %1 : Function = prim::Constant[name="delayed"]()
  %3 : Tensor = prim::CallFunction(%1, %x.1) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:399:17
  return (%3)
with prim::awaitable_then_1 = graph(%aw.5 : Await(Tensor),
      %aw_out : Tensor):
  %3 : Function = prim::Constant[name="then"]()
  %4 : Tensor = prim::CallFunction(%3, %aw.5, %aw_out) # /home/ivankobzarev/github/pytorch/test/jit/test_await.py:401:12
  return (%4)
```

`python_print.cpp` produces:

```
def main(x: Tensor) -> Tensor:
  def __awaitable_function():
    _0 = __torch__.jit.test_await.delayed(x, )
    return _0
  aw = awaitable(__awaitable_function)
  z = __torch__.jit.test_await.gap(x, )
  def __awaitable_then_function(aw0 : Await[Tensor], x0 : Tensor):
    _1 = __torch__.jit.test_await.then(aw0, x0, )
    return _1
  awaitable_then(__awaitable_then_function, aw)
  y = ops.prim.awaitable_wait(aw)
  return torch.add(torch.add(x, y), z)
```



Differential Revision: [D43572964](https://our.internmc.facebook.com/intern/diff/D43572964)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43572964).

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43572964).

davidberard98 commented on torch/csrc/jit/passes/inline_forked_closures.cpp:
```diff
@@ -58,6 +58,88 @@ void inlineForkedClosure(Node* fork_closure, NodeKind genKind) {
+  Node* function = function_context_node->inputs().at(0)->node();
+  Node* context = function_context_node->inputs().at(1)->node();
+  auto then_graph = function->g(attr::Subgraph)->copy();
+  // Replacing ir_emitter emitted awaitableThenClosure, with awaitable_then node
+  // with 2 inputs: Await Value which graph it continues, the fake input Value
+  // which represents the future result of Await graph with previously added
+  // continuations execution. As we do not have Value for it in IR - we add
+  // no-op node awaitable_then_input to produce this Value and keep the graph
+  // valid. At runtime instead the result of Await will be placed for this
```
Thanks a lot for this.

nit: I suggested how to re-phrase this below. (I find this confusing so I think it would be good to be very clear in the comments, so future developers have a better chance of understanding!). Please make sure what I wrote makes sense (and add the part I was wondering about)

> ir_emitter produces IR with prim::awaitableThenClosure; but just like the inlineForkedClosure method above, we need to modify it before running it.
> We replace the prim::Closure with a prim::awaitable_then, which is a callable.
> Note that a prim::awaitable_then function takes arguments (%x: Await(T), %y: T). In the calling graph, we will have %x: Await(T) available; it's the awaitable that the ops are operating on. However, we won't have second argument %y: T. (At execution time, %y, the input to the function we're calling as part of our awaitable_then call, will be the output of the previous function queued in the Awaitable object). As a workaround, we use prim::awaitable_then_input to create a dummy argument in the calling graph that we pass into the awaitable_then, which makes the graph valid.

...
TODO - @IvanKobzarev : can you comment about needing to unpack the context tuple? Will the context tuple ever be non-empty?
...

IvanKobzarev replied:
Thanks a lot for the comment wording, this is historically one of the hardest parts for me :) Changed the comment.

IvanKobzarev committed:
Update on "[jit][await] Add await then fns"

Introducing _awaitable_then operator to add continuation function to the awaitable.

For example:
```
        def delayed(x: Tensor) -> Tensor:
            return -1 * x

        def then(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 5 * aw_out

        def then2(aw: Await[Tensor], aw_out: Tensor) -> Tensor:
            return 4 * aw_out

        def main(x: Tensor) -> Tensor:
            aw = torch.jit._awaitable(delayed, x)
            torch.jit._awaitable_then(then, aw)
            torch.jit._awaitable_then(then2, aw)
            return torch.jit._awaitable_wait(aw)
```
`awaitable_wait(aw)` is `then2(then(delayed(x)))`

Inside ivalue `then` continuations are stored as std::function which will be called on the result when `wait` is called.

IR:

torch.jit._awaitable_then first will emit `awaitableThenClosure` (the same as fork, awaitable), 
but that closure will have two arguments: `Await[T]` and fake argument of `T` type, produced by no-op operator `awaitable_then_input`.
This fake argument in runtime will become the result of Await delayed function.
We need this fake argument to generate valid call to `then` function without having that Value at the moment of generation.

(This happens in `torch/csrc/jit/frontend/ir_emitter.cpp` and `torch/csrc/jit/passes/inline_forked_closures.cpp` where `prim::awaitable_then_input` is added)


New instruction `AWAITABLE_THEN` is added, which adds continuation to Await on the stack (the first argument of Closure).


python_print:

As awaitable_then_input is a fake argument - it is ignored for printing.
awaitable_then is printed exactly the same way as python is specified.

Python:
```
def gap(x: Tensor):
    return torch.relu(x)

def delayed(x: Tensor) -> Tensor:
    return -1 * x

def continuation_fn(aw: Await[Tensor], x: Tensor) -> Tensor:
    return 5 * x

def main(x: Tensor) -> Tensor:
    aw = torch.jit._awaitable(delayed, x)
    z = gap(x)
    torch.jit._awaitable_then(continuation_fn, aw)

    y = torch.jit._awaitable_wait(aw)
    return x + y + z
```

IR flow:

```
graph(%self : __torch__.PlaceholderModule,
      %x.1 : Tensor):
  %41 : Tensor = prim::Uninitialized()
  %40 : bool = prim::Uninitialized()
  %38 : bool = prim::Constant[value=1]()
  %39 : bool = prim::Constant[value=0]()
  %37 : bool = prim::Uninitialized()
  %35 : bool = prim::Constant[value=1]()
  %36 : bool = prim::Constant[value=0]()
  %32 : bool = prim::Constant[value=0]()
  %2 : NoneType = prim::Closure_0()
  %42 : (Tensor) = prim::TupleConstruct(%x.1)
  %43 : (NoneType, (Tensor)) = prim::TupleConstruct(%2, %42)
  %aw.1 : Await(Tensor) = prim::awaitableClosure(%43)
  %10 : Function = prim::Constant[name="gap"]()
  %z.1 : Tensor = prim::CallFunction(%10, %x.1)
  %12 : NoneType = prim::Closure_1()
  %44 : () = prim::TupleConstruct()
  %45 : (NoneType, ()) = prim::TupleConstruct(%12, %44)
   = prim::awaitableThenClosure(%45, %aw.1)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1)
  %26 : int = prim::Constant[value=1]()
  %27 : Tensor = aten::add(%x.1, %y.1, %26)
  %29 : int = prim::Constant[value=1]()
  %30 : Tensor = aten::add(%27, %z.1, %29)
  return (%30)
with prim::Closure_0 = graph(%context : (Tensor)):
  %2 : Function = prim::Constant[name="delayed"]()
  %x.1 : Tensor = prim::TupleUnpack(%context)
  %3 : Tensor = prim::CallFunction(%2, %x.1) # :0:0
  return (%3)
with prim::Closure_1 = graph(%context : (),
      %aw0.1 : Await(Tensor),
      %x0.1 : Tensor):
  %4 : Function = prim::Constant[name="continuation_fn"]()
  %5 : Tensor = prim::CallFunction(%4, %aw0.1, %x0.1) # :0:0
  return (%5)
```
After applying `inline_forked_closures.cpp`
```
graph(%x.1 : Tensor):
  %19 : int = prim::Constant[value=1]()
  %7 : Function = prim::Constant[name="gap"]()
  %aw.1 : Await(Tensor) = prim::awaitable_0(%x.1)
  %z.1 : Tensor = prim::CallFunction(%7, %x.1)
  %40 : Tensor = prim::awaitable_then_input(%aw.1)
   = prim::awaitable_then_1(%aw.1, %40)
  %y.1 : Tensor = prim::awaitable_wait(%aw.1)
  %20 : Tensor = aten::add(%x.1, %y.1, %19)
  %23 : Tensor = aten::add(%20, %z.1, %19)
  return (%23)
with prim::awaitable_0 = graph(%x.1 : Tensor):
  %1 : Function = prim::Constant[name="delayed"]()
  %3 : Tensor = prim::CallFunction(%1, %x.1)
  return (%3)
with prim::awaitable_then_1 = graph(%aw.5 : Await(Tensor),
      %aw_out : Tensor):
  %3 : Function = prim::Constant[name="continuation_fn"]()
  %4 : Tensor = prim::CallFunction(%3, %aw.5, %aw_out)
  return (%4)
```

`python_print.cpp` produces:

```
def main(x: Tensor) -> Tensor:
  def __awaitable_function():
    _0 = __torch__.jit.test_await.delayed(x, )
    return _0
  aw = awaitable(__awaitable_function)
  z = __torch__.jit.test_await.gap(x, )
  def __awaitable_then_function(aw0 : Await[Tensor], x0 : Tensor):
    _1 = __torch__.jit.test_await.continuation_fn(aw0, x0, )
    return _1
  awaitable_then(__awaitable_then_function, aw)
  y = ops.prim.awaitable_wait(aw)
  return torch.add(torch.add(x, y), z)
```



Differential Revision: [D43572964](https://our.internmc.facebook.com/intern/diff/D43572964)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43572964).

IvanKobzarev commented:
Renamed `then -> continuation_fn` everywhere (PR description, comment, test)
